bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationClient.java

```diff
@@ -447,4 +461,46 @@ public synchronized void unwatchReadOnlyBookies(RegistrationListener listener) {
         return newBookieAddrs;
     }
 
+    private static BookieId stripBookieIdFromPath(String path) {
+        final int slash = path.lastIndexOf('/');
+        if (slash >= 0) {
+            try {
+                return BookieId.parse(path.substring(slash + 1));
+            } catch (IllegalArgumentException e) {
+                log.warn("Cannot decode bookieId from {}", path, e);
+            }
+        }
+        return null;
+    }
+
+    private class BookieServiceInfoCacheInvalidationWatcher implements Watcher {
+
```

Comment:
How frequently can a bookieID change? What factors drive its change?

Reply:
I see these known cases:
- a bookie with a dynamic network address
- changing configuration (change DNS name....)
- you change configuration and add new endpoints (this will happen in the future, when for instance @sijie will contribute the pure TLS endpoint)

So basically I think they won't change very ofter